### PR TITLE
Introduce SecureMiningReward action

### PR DESCRIPTION
**Pull request by longfin**
This PR introduces `SecureMiningReward` to retrieve NCG from authorized miners to arbitrary addresses, for [NCIP-10](https://docs.google.com/document/d/1ErZ5JQia03KqXRG6IRZ7SORfnxMLZfJg4patVKFGX5Y/edit).

### CODE REVIEW

File: Lib9c/Action/SecureMiningReward.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+            {
+                return state.MarkBalanceChanged(
+                    Amount.Currency,
+                    new [] { Miner, Recipient}
+                );
+            }
+
+            CheckPermission(context);
+
+            // TODO consider a better solution to determine.
```
Comment (longfin) [inline Lib9c/Action/SecureMiningReward.cs:47]:
```suggestion
            // FIXME considers better solutions to check.
```

File: Lib9c/Action/SecureMiningReward.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+            if (!BlockChain.Policy.BlockPolicySource.AuthorizedMiners.Contains(Miner))
+            {
+                throw new InvalidAddressException($"Given address[{Miner}] isn't an authorized miner");
+            }
+
+            // TODO Do we need the maximum cap per miner? (but... if then, how do we maintain?)
```
Comment (longfin) [inline Lib9c/Action/SecureMiningReward.cs:53]:
```suggestion
```

File: Lib9c/Action/SecureMiningReward.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        );
+
+        public override IAccountStateDelta Execute(IActionContext context)
+        {
+            IAccountStateDelta state = context.PreviousStates;
+            if (context.Rehearsal)
+            {
+                return state.MarkBalanceChanged(
+                    Amount.Currency,
+                    new [] { Miner, Recipient}
```
Comment (moreal) [inline Lib9c/Action/SecureMiningReward.cs:51]: Just style comment, you can ignore.

```suggestion
                    new [] { Miner, Recipient }
```

**Comment by moreal** [comment]
Well, as I understood, the owner according to `AdminState` will control the assets of authorized miners. And according to NCIP-10, I understood the treasury (maybe Planetarium) will get 40%, 40% will be burnt and 20% must be retained. But the code seems not to handle anything about the rule. Is it based on the assumption that *Planetarium* can do it well? Or just not considered?

**Review by moreal — COMMENTED** [review]
Well, as I understood, the owner according to `AdminState` will control the assets of authorized miners. And according to NCIP-10, I understood the treasury (maybe Planetarium) will get 40%, 40% will be burnt and 20% must be retained. But the code seems not to handle anything about the rule. Is it based on the assumption that *Planetarium* can do it well? Or just not considered?

**Comment by longfin** [comment]
> Well, as I understood, the owner according to `AdminState` will control the assets of authorized miners. And according to NCIP-10, I understood the treasury (maybe Planetarium) will get 40%, 40% will be burnt and 20% must be retained. But the code seems not to handle anything about the rule. Is it based on the assumption that _Planetarium_ can do it well? Or just not considered?

It is assumed that the user who is authorized (i.e., passes `CheckPermission()`) performs "well". 😅

**Comment by longfin** [comment]
> > It is assumed that the user who is authorized (i.e., passes `CheckPermission()`) performs "well". 😅
> 
> I'm very careful to reply. But, in my honest opinion, I think it is better to automate by implementing the NCIP-10 as the action code, to block the possibilities of mistakes. It seems able to be implemented.
> 
> Of course, it is just my preferred way so if they prefer this way, I will not object.

@moreal First, thanks for your opinion and I also understand that this action seems quite strange.

- Despiting I've mentioned NCIP-10, it lacks some information to apply directly (e.g., it doesn't specify the actual starting block or miner address).
   - but on second thought, it had an address to withdraw (i.e., `0xB3bCa3b3c6069EF5Bdd6384bAD98F11378Dc360E`). so it might be better to force that address.
- Even if we fixed block indices arbitrarily, it isn't matched the actual amount for a specific miner because we've changed the addresses of the miner for reasons such as outages.

To sum up, if we secure all miner's rewards for all periods instead of NCIP-10's, it can be executed automatically and easily.

@jckdotim do you have any idea?

**Comment by longfin** [comment]
Amended https://github.com/planetarium/lib9c/pull/1720/commits/2d2d018a7e0d057168b638c77f0ba3013038e1b3  to address https://github.com/planetarium/lib9c/pull/1720#issuecomment-1439915335. 

PTAL @moreal @jckdotim

**Comment by jckdotim** [comment]
@longfin I agree with your suggestion to automate the process as the action code to prevent mistakes.

**Review by moreal — APPROVED** [review]
